File: cmd/snapshot.go
package cmd

import (
	"fmt"
	"os"

	"github.com/CraftIntertech/craft-cli/internal/client"
	"github.com/CraftIntertech/craft-cli/internal/output"
	"github.com/spf13/cobra"
)

func init() {
	// snapshot list
	snapshotListCmd := &cobra.Command{
		Use:   "list [VM_ID]",
		Short: "List snapshots for a VM",
		Args:  cobra.MaximumNArgs(1),
		Run:   snapshotListRun,
	}

	// snapshot create
	snapshotCreateCmd := &cobra.Command{
		Use:   "create [VM_ID]",
		Short: "Create a snapshot for a VM",
		Args:  cobra.MaximumNArgs(1),
		Run:   snapshotCreateRun,
	}
	snapshotCreateCmd.Flags().String("description", "", "Snapshot description")

	// snapshot delete
	snapshotDeleteCmd := &cobra.Command{
		Use:   "delete [VM_ID] [SNAP_ID]",
		Short: "Delete a snapshot",
		Args:  cobra.MaximumNArgs(2),
		Run:   snapshotDeleteRun,
	}

	// snapshot rollback
	snapshotRollbackCmd := &cobra.Command{
		Use:   "rollback [VM_ID] [SNAP_ID]",
		Short: "Rollback a VM to a snapshot",
		Args:  cobra.MaximumNArgs(2),
		Run:   snapshotRollbackRun,
	}

	// snapshot sync
	snapshotSyncCmd := &cobra.Command{
		Use:   "sync [VM_ID]",
		Short: "Sync snapshots for a VM",
		Args:  cobra.MaximumNArgs(1),
		Run:   snapshotSyncRun,
	}

	SnapshotCmd.AddCommand(snapshotListCmd)
	SnapshotCmd.AddCommand(snapshotCreateCmd)
	SnapshotCmd.AddCommand(snapshotDeleteCmd)
	SnapshotCmd.AddCommand(snapshotRollbackCmd)
	SnapshotCmd.AddCommand(snapshotSyncCmd)
}

func snapshotListRun(cmd *cobra.Command, args []string) {
	vmID := ""
	if len(args) > 0 {
		vmID = args[0]
	} else {
		vmID = SelectVM("Select a VM:")
	}

	path := fmt.Sprintf("/vms/%s/snapshots", vmID)
	result := client.Get(path, nil, true)

	data, ok := result["data"].([]interface{})
	if !ok {
		fmt.Fprintln(os.Stderr, "No snapshots found.")
		return
	}

	var rows [][]string
	for _, item := range data {
		snap, ok := item.(map[string]interface{})
		if !ok {
			continue
		}
		rows = append(rows, []string{
			fmt.Sprintf("%v", snap["id"]),
			fmt.Sprintf("%v", snap["description"]),
			fmt.Sprintf("%v", snap["created"]),
		})
	}

	output.PrintTable(rows, []string{"ID", "Description", "Created"})
}

func snapshotCreateRun(cmd *cobra.Command, args []string) {
	vmID := ""
	if len(args) > 0 {
		vmID = args[0]
	} else {
		vmID = SelectVM("Select a VM:")
	}

	description, _ := cmd.Flags().GetString("description")

	body := map[string]interface{}{}
	if description != "" {
		body["description"] = description
	}

	path := fmt.Sprintf("/vms/%s/snapshots", vmID)
	client.Post(path, body, true, 0)

	output.PrintSuccess("Snapshot created successfully.")
}

func snapshotDeleteRun(cmd *cobra.Command, args []string) {
	vmID := ""
	snapID := ""

	switch len(args) {
	case 2:
		vmID = args[0]
		snapID = args[1]
	case 1:
		vmID = args[0]
		snapID = PromptText("Enter snapshot ID to delete:")
	default:
		vmID = SelectVM("Select a VM:")
		snapID = PromptText("Enter snapshot ID to delete:")
	}

	if !Confirm(fmt.Sprintf("Delete snapshot %s?", snapID)) {
		output.PrintInfo("Cancelled.")
		return
	}

	path := fmt.Sprintf("/vms/%s/snapshots/%s", vmID, snapID)
	client.Delete(path, true)

	output.PrintSuccess("Snapshot deleted successfully.")
}

func snapshotRollbackRun(cmd *cobra.Command, args []string) {
	vmID := ""
	snapID := ""

	switch len(args) {
	case 2:
		vmID = args[0]
		snapID = args[1]
	case 1:
		vmID = args[0]
		snapID = PromptText("Enter snapshot ID to rollback to:")
	default:
		vmID = SelectVM("Select a VM:")
		snapID = PromptText("Enter snapshot ID to rollback to:")
	}

	if !Confirm(fmt.Sprintf("Rollback VM %s to snapshot %s? This may cause data loss.", vmID, snapID)) {
		output.PrintInfo("Cancelled.")
		return
	}

	path := fmt.Sprintf("/vms/%s/snapshots/%s/rollback", vmID, snapID)
	client.Post(path, nil, true, 0)

	output.PrintSuccess("Snapshot rollback initiated successfully.")
}

func snapshotSyncRun(cmd *cobra.Command, args []string) {
	vmID := ""
	if len(args) > 0 {
		vmID = args[0]
	} else {
		vmID = SelectVM("Select a VM:")
	}

	path := fmt.Sprintf("/vms/%s/snapshots/sync", vmID)
	client.Post(path, nil, true, 0)

	output.PrintSuccess("Snapshot sync initiated successfully.")
}

File: craft/commands/snapshot.py
import click

from craft.client import delete, get, post
from craft.output import print_item, print_json, print_success, print_table


@click.command("list")
@click.argument("vm_id", required=False, default=None)
def snap_list(vm_id):
    """List snapshots for a VM."""
    if not vm_id:
        from craft.interactive import select_vm
        vm_id = select_vm()
    data = get(f"/vms/{vm_id}/snapshots")
    items = data.get("data", data)
    if isinstance(items, dict):
        items = items.get("snapshots", items.get("items", []))
    if isinstance(items, list) and items:
        rows = []
        for s in items:
            rows.append([
                s.get("id", s.get("name", "")),
                s.get("description", ""),
                s.get("createdAt", s.get("snaptime", "")),
            ])
        print_table(rows, ["ID", "Description", "Created"])
    else:
        print_json(data)


@click.command("create")
@click.argument("vm_id", required=False, default=None)
@click.option("--description", default="", help="Snapshot description (max 200 chars)")
def snap_create(vm_id, description):
    """Create a snapshot (max 5 per VM)."""
    if not vm_id:
        from craft.interactive import select_vm
        vm_id = select_vm()
    if not description:
        from craft.interactive import input_text
        description = input_text("Description (optional):")
    body = {}
    if description:
        body["description"] = description
    data = post(f"/vms/{vm_id}/snapshots", body)
    print_success("Snapshot created.")
    print_item(data)


@click.command("delete")
@click.argument("vm_id", required=False, default=None)
@click.argument("snap_id", required=False, default=None)
def snap_delete(vm_id, snap_id):
    """Delete a snapshot."""
    if not vm_id:
        from craft.interactive import select_vm
        vm_id = select_vm()
    if not snap_id:
        from craft.interactive import select_snapshot
        snap_id = select_snapshot(vm_id, "Select snapshot to delete")
    if not click.confirm("Delete this snapshot?"):
        click.echo("Cancelled.")
        return
    delete(f"/vms/{vm_id}/snapshots/{snap_id}")
    print_success(f"Snapshot {snap_id} deleted.")


@click.command("rollback")
@click.argument("vm_id", required=False, default=None)
@click.argument("snap_id", required=False, default=None)
def snap_rollback(vm_id, snap_id):
    """Rollback VM to a snapshot."""
    if not vm_id:
        from craft.interactive import select_vm
        vm_id = select_vm()
    if not snap_id:
        from craft.interactive import select_snapshot
        snap_id = select_snapshot(vm_id, "Select snapshot to rollback to")
    if not click.confirm("WARNING: Current VM state will be lost. Continue?"):
        click.echo("Cancelled.")
        return
    post(f"/vms/{vm_id}/snapshots/{snap_id}/rollback")
    print_success("Rollback complete.")


@click.command("sync")
@click.argument("vm_id", required=False, default=None)
def snap_sync(vm_id):
    """Sync snapshot records with hypervisor."""
    if not vm_id:
        from craft.interactive import select_vm
        vm_id = select_vm()
    post(f"/vms/{vm_id}/snapshots/sync")
    print_success("Snapshots synced.")
